src/man/man3/MonitorCorner.3
.TH MonitorCorner 3 "29 May 2002" "HulkRobot" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MonitorCorner \- 
.SH SYNOPSIS
.br
.PP
\fC#include <MonitorCorner.h>\fR
.PP
.SS Public Methods

.in +1c
.ti -1c
.RI "\fBMonitorCorner\fR (int priority, int samplingPeriod, int stackSize, \fBgyro\fR *aG, \fBusonic\fR *aU, \fBwheel\fR *aW)"
.br
.ti -1c
.RI "\fB~MonitorCorner\fR ()"
.br
.ti -1c
.RI "int \fBinit\fR (void)"
.br
.ti -1c
.RI "int \fBstep\fR (void)"
.br
.ti -1c
.RI "int \fBrun\fR (void)"
.br
.ti -1c
.RI "int \fBstop\fR (void)"
.br
.ti -1c
.RI "int \fBisCorner\fR (void)"
.br
.in -1c
.SH CONSTRUCTOR & DESTRUCTOR DOCUMENTATION
.PP 
.SS MonitorCorner::MonitorCorner (int _priority, int _samplingPeriod, int stackSize, \fBgyro\fR * aG, \fBusonic\fR * aU, \fBwheel\fR * aW)
.PP
Definition at line 19 of file MonitorCorner.cc.
.PP
.nf
00021 {
00022 
00023   Error( ErrorDebug, "MonitorDooCorner initializing");
00024   priority        = _priority;
00025   samplingPeriod= _samplingPeriod;
00026   delayTicks      = samplingPeriod/OSTickPeriod;
00027 
00028   taskMode        = TaskStopped;
00029   taskSem = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00030   tickerSem       = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00031   taskExitSem     = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00032   tickerExitSem   = semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
00033 
00034   aGyro = aG;
00035   aUSonic = aU;
00036   aWheel = aW;
00037 
00038   Assert( taskSem && taskExitSem && tickerSem && tickerExitSem );
00039 
00040   taskId  = taskSpawn("MonitorCorner", priority, VX_FP_TASK, 
00041                           stackSize, (FUNCPTR) taskModuleStub,
00042                           (int) this, 0, 0, 0, 0, 0, 0, 0, 0, 0);
00043   Assert(taskId!=ERROR);
00044 
00045   semTake(taskExitSem, WAIT_FOREVER);
00046   
00047   tickerTaskId    = taskSpawn("MonitorCorner", TickerTaskPriority, 0, 
00048                           TickerTaskStackSize,
00049                           (FUNCPTR) taskModuleTickerStub,
00050                           (int) this,
00051                           0, 0, 0, 0,
00052                           0, 0, 0, 0, 0);
00053   Assert(tickerTaskId!=ERROR);
00054   semTake(tickerExitSem, WAIT_FOREVER);
00055 
00056   corner = 0; // set initial monitor signal ,false
00057 
.fi
.SS MonitorCorner::~MonitorCorner ()
.PP
Definition at line 59 of file MonitorCorner.cc.
.PP
.nf
00060 {
00061   taskMode=TaskTerminate;
00062   semGive(tickerSem);
00063   semGive(taskSem);
00064   semTake(tickerExitSem, WAIT_FOREVER);
00065   semTake(taskExitSem, WAIT_FOREVER);
00066   semDelete(taskSem);
00067   semDelete(taskExitSem);
00068   semDelete(tickerSem);
00069   semDelete(tickerExitSem);
00070 
00071   Error( ErrorDebug, "MonitorCorner destroyed");
00072 
.fi
.SH MEMBER FUNCTION DOCUMENTATION
.PP 
.SS int MonitorCorner::init (void)
.PP
Definition at line 173 of file MonitorCorner.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00174 {
00175   corner = 0;     // set initially Corner = false
00176 
.fi
.SS int MonitorCorner::isCorner (void)
.PP
Definition at line 167 of file MonitorCorner.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00168 {
00169   return corner;
00170 
.fi
.SS int MonitorCorner::run (void)
.PP
Definition at line 185 of file MonitorCorner.cc.
.PP
Referenced by hulkmaincpp().
.PP
.nf
00186 {
00187   taskMode=TaskRunning;
00188   semGive(tickerSem);
00189   return 0;
00190 
.fi
.SS int MonitorCorner::step (void)
.PP
Definition at line 178 of file MonitorCorner.cc.
.PP
.nf
00179 {
00180   taskMode=TaskStep;
00181   semGive(taskSem);
00182   return 0;
00183 
.fi
.SS int MonitorCorner::stop (void)
.PP
Definition at line 192 of file MonitorCorner.cc.
.PP
.nf
00193 {
00194   taskMode=TaskStopped;
00195   return 0;
00196 
.fi


.SH AUTHOR
.PP 
Generated automatically by Doxygen for HulkRobot from the source code.

src/monitor/MonitorCorner.cc
//
//	File MonitorCorner.cc
//
//	Monitor Corner module for avoid obstacle
//
//	Framework is originally written by Yunho Jeon
//
//	Written by Insop Song
//	1997 12 1
//	ver 0.1
//

#include "dfl.h"
#include "MonitorCorner.h"

unsigned MonitorCorner::range[16];
unsigned MonitorCorner::detected[8];

MonitorCorner::MonitorCorner(int _priority, int _samplingPeriod, 
		int stackSize, gyro *aG, usonic *aU, wheel *aW)
{

  Error( ErrorDebug, "MonitorDooCorner initializing");
  priority 	= _priority;
  samplingPeriod= _samplingPeriod;
  delayTicks	= samplingPeriod/OSTickPeriod;

  taskMode	= TaskStopped;
  taskSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  tickerSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  taskExitSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);
  tickerExitSem	= semBCreate(SEM_Q_PRIORITY, SEM_EMPTY);

  aGyro = aG;
  aUSonic = aU;
  aWheel = aW;

  Assert( taskSem && taskExitSem && tickerSem && tickerExitSem );

  taskId 	= taskSpawn("MonitorCorner", priority, VX_FP_TASK, 
				stackSize, (FUNCPTR) taskModuleStub,
				(int) this, 0, 0, 0, 0, 0, 0, 0, 0, 0);
  Assert(taskId!=ERROR);

  semTake(taskExitSem, WAIT_FOREVER);
  
  tickerTaskId	= taskSpawn("MonitorCorner", TickerTaskPriority, 0, 
				TickerTaskStackSize,
				(FUNCPTR) taskModuleTickerStub,
				(int) this,
				0, 0, 0, 0,
				0, 0, 0, 0, 0);
  Assert(tickerTaskId!=ERROR);
  semTake(tickerExitSem, WAIT_FOREVER);

  corner = 0; // set initial monitor signal ,false
}

MonitorCorner::~MonitorCorner()
{
  taskMode=TaskTerminate;
  semGive(tickerSem);
  semGive(taskSem);
  semTake(tickerExitSem, WAIT_FOREVER);
  semTake(taskExitSem, WAIT_FOREVER);
  semDelete(taskSem);
  semDelete(taskExitSem);
  semDelete(tickerSem);
  semDelete(tickerExitSem);

  Error( ErrorDebug, "MonitorCorner destroyed");
}

//
// C stub functions to call C++ member functions from
// a newly spawned task
//

void MonitorCorner::taskModuleStub(MonitorCorner *This, int d2, int d3, 
					int d4, int d5, int d6, int d7, 
					int d8, int d9, int d10)
{
  Error( ErrorDebug, "MonitorCorner taskModuleStub ");
  This->task();
}

void MonitorCorner::taskModuleTickerStub(MonitorCorner *This, int d2, 
			      int d3, int d4, int d5, int d6, int d7, 
			      int d8, int d9, int d10)
{
  Error( ErrorDebug, "MonitorCorner taskModuleTickerStub");
  This->ticker();
}

void MonitorCorner::ticker(void)
{
  semGive(tickerExitSem);
  while(taskMode!=TaskTerminate) {

    if(taskMode==TaskStopped || taskMode==TaskPaused) 
       semTake(tickerSem, WAIT_FOREVER);

    while(taskMode==TaskRunning) {
      Error( ErrorNone, "ticker::semGive" );
      semGive(taskSem);
      taskDelay(delayTicks);
    }
  }
  semGive(tickerExitSem);
  // task exit - deleted automatically
}

void MonitorCorner::task(void)
{
  semGive(taskExitSem); 		// signal that it has been created
  while(taskMode!=TaskTerminate) {
    semTake(taskSem, WAIT_FOREVER);	// wait until ticker ticks
    if (taskMode!=TaskRunning && taskMode!=TaskStep) continue;
    Error( ErrorNone, "task tick" );

    // task code HERE
    for(int i=0;i<16;i++) {
      range[i] = aUSonic->usData(i);
    }
    Error(ErrorDebug,"MonitorCorner");

    //
    // check Corner all side sensors , 
    //
    // configuration  [9]-[10]-[11]-[12]-[13]-[14]-[15] 
    //

    if(corner == 0) {	// Corner start check
      if(range[12]>upTHRESHOLDcorner) {
	detectedNumLower = 0;
	detectedNumUpper = 0;
	for(int i=9;i<12;i++) {
	  if(range[i]<THRESHOLDcorner) {
	    detectedNumLower++;
	  }
	}
	for( i=13;i<16;i++) {
	  if(range[i]<THRESHOLDcorner) {
	    detectedNumUpper++;
	  }
	}

	Error2(ErrorWarning," %d  %d ", detectedNumLower, detectedNumUpper);

	if((detectedNumLower >=2)&&(detectedNumUpper <= 1)) {
	  corner = 1;
	  Error( ErrorDebug, "Corner detected" );
	}

      } 
    } else if(corner ==1){		// Corner end check
      if(range[13]<endCorner) {
	corner = 0;
	Error( ErrorDebug, "Corner End" );
      }
    }
  }
  semGive(taskExitSem); 
  // task exit - deleted automatically
}

int MonitorCorner::isCorner(void)
{
  return corner;
}
  
    
int MonitorCorner::init(void)
{
  corner = 0;	// set initially Corner = false
}

int MonitorCorner::step(void)
{
  taskMode=TaskStep;
  semGive(taskSem);
  return 0;
}

int MonitorCorner::run(void)
{
  taskMode=TaskRunning;
  semGive(tickerSem);
  return 0;
}

int MonitorCorner::stop(void)
{
  taskMode=TaskStopped;
  return 0;
}

